src/ML-PACE/pair_pace_al.cpp

```diff
@@ -0,0 +1,599 @@
+                           n);
+        }
+    }
+
+
+    // read args that map atom types to PACE elements
+    // map[i] = which element the Ith atom type is, -1 if not mapped
+    // map[0] is not used
+    delete aceimpl->ace;
+    delete aceimpl->rec_ace;
+
+    aceimpl->ace = new ACEBEvaluator();
+    aceimpl->ace->element_type_mapping.init(atom->ntypes + 1);
+
+    aceimpl->rec_ace = new ACERecursiveEvaluator();
+    aceimpl->rec_ace->set_recursive(true);
+    aceimpl->rec_ace->element_type_mapping.init(atom->ntypes + 1);
+    aceimpl->rec_ace->element_type_mapping.fill(-1); //-1 means atom not included into potential
+
+    FILE *species_type_file = nullptr;
+    if (is_dump_extrapolative_structures)
+        species_type_file = fopen(SPECIES_TYPE_FNAME, "w");
+
+    const int n = atom->ntypes;
+    for (int i = 1; i <= n; i++) {
+        char *elemname = elemtypes[i - 1];
+        if (strcmp(elemname, "NULL") == 0) {
+            // species_type=-1 value will not reach ACE Evaluator::compute_atom,
+            // but if it will ,then error will be thrown there
+            aceimpl->ace->element_type_mapping(i) = -1;
+            map[i] = -1;
+            if (comm->me == 0) utils::logmesg(lmp, "Skipping LAMMPS atom type #{}(NULL)\n", i);
+        } else {
+            // dump species types for reconstruction of atomic configurations
+            if (is_dump_extrapolative_structures) fprintf(species_type_file, "%s ", elemname);
+            int atomic_number = AtomicNumberByName_pace_al(elemname);
+            if (atomic_number == -1) error->all(FLERR, "'{}' is not a valid element\n", elemname);
+            SPECIES_TYPE mu = aceimpl->basis_set->get_species_index_by_name(elemname);
+            if (mu != -1) {
+                if (comm->me == 0)
+                    utils::logmesg(lmp, "Mapping LAMMPS atom type #{}({}) -> ACE species type #{}\n", i,
+                                   elemname, mu);
+                map[i] = mu;
+                // set up LAMMPS atom type to ACE species  mapping for ace evaluators
+                aceimpl->ace->element_type_mapping(i) = mu;
+                aceimpl->rec_ace->element_type_mapping(i) = mu;
+            } else {
+                error->all(FLERR, "Element {} is not supported by ACE-potential from file {}", elemname,
+                           potential_file_name);
+            }
+        }
+    }
+    if (is_dump_extrapolative_structures) fclose(species_type_file);
+    aceimpl->ace->set_basis(*aceimpl->basis_set);
+    aceimpl->rec_ace->set_basis(*aceimpl->ctilde_basis_set);
+
+    if (comm->me == 0) utils::logmesg(lmp, "Loading ASI {}\n", active_set_inv_filename);
+    aceimpl->ace->load_active_set(active_set_inv_filename);
+
+    // clear setflag since coeff() called once with I,J = * *
+    for (int i = 1; i <= n; i++)
+        for (int j = i; j <= n; j++)
+            scale[i][j] = 1.0;
+
+    // prepare compute gamma
+    if (!computePaceAtom) {
```

Comment:
Dear Axel, thank you for all your comments I will try to implement them. 
Regarding the dump created by pair_styl PaceActiveLearning: main pecularity of this solution is that the dump action will be invoked **non-regularly** by pair_style::compute [here](https://github.com/yury-lysogorskiy/lammps/blob/feature/pace-al/src/ML-PACE/pair_pace_al.cpp#L339) only if certain condition is met (i.e. extrapolation grade exceed the threshold). 
But as far as I understand, the user-created dump will be called **regularly** every N steps (defined by user), which is not what we need...

As for the compute style - I agree, it could/should be created by user, BUT, the  before-mentioned dump should use this compute per-atom variable as well, effectively:
```
dump pace all custom freq extrapolation.dat id type mass x y z c_pace_gamma
```

So my logic was that there are dependencies chains: 
1) dump should be called non-regularly -> not register dump dump by lmp object -> create it manually in pair::coeff 
2) dump requires per-atom variable c_pace_gamma -> need compute pace/al  ->  a) if compute is created by user after pair::coeff,  dump will in pair::coeff will crush; b) if created by user before setting up a dump , then compute will not find pair_style pace/al. So compute pace/al should be created in pace::coeff  BEFORE creating dump.

if we could call  user-defined (also *registered* in lammps) dump NON-REGULARLY, then it will solve the dependencies...

Reply:
Dump output at irregular intervals is possible. There is `dump_modify every` which accepts a variable. Please checkout the dump_modify docs.
One possible way to deal with the situation could be to create a dump style "pace" or "pace/al" that is derived from dump style custom and then has extra checks included and automatically synchronizes its output with information from the pair style.